polkadot/runtime/common/src/slots/mod.rs

```diff
@@ -267,12 +350,19 @@ impl<T: Config> Pallet<T> {
 				if let Some(ended_lease) = maybe_ended_lease {
 					// Then we need to get the new amount that should continue to be held on
 					// deposit for the parachain.
-					let now_held = Self::deposit_held(para, &ended_lease.0);
+					let required_hold = Self::required_deposit(para, &ended_lease.0);
 
 					// If this is less than what we were holding for this leaser's now-ended lease,
 					// then unreserve it.
-					if let Some(rebate) = ended_lease.1.checked_sub(&now_held) {
-						T::Currency::unreserve(&ended_lease.0, rebate);
+					let current_hold = Self::deposit_held(para, &ended_lease.0);
+
+					// Note: we check reserved balance from state instead of upcoming leases since
+					// we might have unreserved early. Since reserves are not named, we can't check
+					// if this reserve was made by slots pallet or not. An account could probably
+					// cheat by reserving amount for something else so better to track reserved
+					// amount at all times.
+					if let Some(rebate) = current_hold.checked_sub(&required_hold) {
+						Self::unreserve(para, &ended_lease.0, rebate);
 					}
```

Comment:
nit but how about an `else` arm here with defensive?

Reply:
It is possible for current hold to be less than required hold when there is an early lease refund. But actually it should never be possible at a start of a lease.
Added a defensive check 👍.